### PR TITLE
Make reachability cache fully consistent with map::sees, fix mirrors

### LINKED ISSUE #46022: Peeking through curtain not working - Build #11250
# Peeking through curtain not working as of Build #11250 

Using peek through curtain doesn't show any mobs behind the window

# Steps To Reproduce

1. Make character.
2. Use peek through curtain option at window with closed curtains with mob behind window.
3. Mob will not show either visualy or in the mob list.

# Expected behavior

Mob should show either visualy or in the mob list.

# Screenshots

obrazek

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 1909
- Game Version: 0.E-8138-g3b97eed [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions],
    Stats Through Skills [StatsThroughSkills]
]

# Additional context

### PR DESCRIPTION
**Pull request by Aivean**
#### Summary

SUMMARY: Bugfixes "Make reachability cache fully consistent with map::sees, fix mirrors"

#### Purpose of change

An alternative to #46046, fixes #46022.

Currently reachability cache falsely rejects queries when the source or target tile is opaque. In-game that could be noticeable when player is peeking through the opaque curtain (#46022).

Additionally, currently reachability cache is used for player's vision, which breaks mirrors, as mirror los is different from `map::sees`.

#### Describe the solution

1. Adjusted DP function in reachability cache, such that if source and/or destination tiles are opaque this still counts as a valid LOS.  Added unit test to verify consistency between `map::sees` and `map::has_potential_los`.

2. Removed `map::has_potential_los` from `Creature:sees` when creature `is_player` to fix mirrors.

#### Describe alternatives you've considered

A temporary fix was proposed in #46046.

#### Testing

Added unit test for reachability cache.

<details>
<summary>Before the change in dynamic function is was failing as expected</summary>

```
../tests/reachability_cache_test.cpp:65: FAILED:
  CHECK( ( !map_sees || r_cache_sees ) )
with expansion:
  false
with messages:
  .............
  .............
  ......####...
  ......#..#F..
  ...####..#...
  ...#..X..T...
  ...#.....#...
  ...#..#..#...
  ...#######...
  .............
  .............
  src := (64,58,0)
  dest := (63,60,0)
  map_sees := true
  r_cache_sees := false


../tests/reachability_cache_test.cpp:65: FAILED:
  CHECK( ( !map_sees || r_cache_sees ) )
with expansion:
  false
with messages:
  .............
  .............
  ......####...
  ......#.F#...
  ...####..#...
  ...#..X..#...
  ...#.....#...
  ...#..#..#...
  ...####T##...
  .............
  .............
  src := (62,58,0)
  dest := (61,63,0)
  map_sees := true
  r_cache_sees := false


//vertical
../tests/reachability_cache_test.cpp:62: FAILED:
  CHECK( ( !map_sees || r_cache_sees ) )
with expansion:
  false
with messages:
  F — from, T — to
  .............
  .............
  ......####...
  ......#^^#...
  ...####..#...
  ...#..X..#...
  ...#...^^#...
  ...#..#^^#...
  ...######F...
  .......^^^...
  ...........T.
  src := (63,63,0)
  dest := (65,65,1)
  map_sees := true
  r_cache_sees := false
```

</details>

Playtested different scenarios, visually evaluated that reachability cache is still working.

Peeking:
peek

Mirrors:
(blue tinted tiles are rejected by reachability cache, but monster is still seen)
image

### CODE REVIEW

File: tests/reachability_cache_test.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            CAPTURE( src, dest, map_sees, r_cache_sees );
+            // when map_sees==true, r_cache_sees should be true,
+            // but when !map_sees, r_cache_sees is allowed be false or true
+            CHECK( ( !map_sees || r_cache_sees ) );
+        }
+    }
+
+    // at least some lookups should be rejected by reachability cache
+    CHECK( rejected_cnt > 100 );
+    WARN( "rejected count: " << rejected_cnt );
```
Comment (kevingranade) [inline tests/reachability_cache_test.cpp:68]: This is a bit spammy in practice, is this leftover from debugging?

Comment (Aivean) [reply tests/reachability_cache_test.cpp:68]: I left it there intentionally, but now that I'm thinking about it, there is no real reason for it to be there.

Comment (kevingranade) [reply tests/reachability_cache_test.cpp:68]: It's especially visible if you run cata_tests with no args, which normally prints out a 5 line summary if nothing broke.

Comment (kevingranade) [inline tests/reachability_cache_test.cpp:68]:
```suggestion
    // at least some lookups should be rejected by reachability cache
    CHECK( rejected_cnt > 100 );
```

**Pull request merged**